content/posts/maths/NMPC-Two-Wheel-Bicycle-Model.md

```diff
@@ -0,0 +1,186 @@
+
+<p>
+$$
+\begin{aligned}
+&\dot{p}_x = v_x \cos{\psi} - v_y \cos{\psi}, \\
+&\dot{p}_y = v_x \sin{\psi} + v_y \cos{\psi}, \\
+&\dot{\psi} = \omega, \\
+&m \dot{v}_x = F_{r, x} - F_{f, y} \sin{\delta} - F_{f, x} \cos{\delta} + m v_y \omega, \\
+&m \dot{v}_y = F_{r, y} + F_{f, y} \cos{\delta} + F_{f, x} \sin{\delta} - m v_x \omega, \\
+&J_z \dot{\omega} = l_f F_{f, y} \cos{\delta} + l_f F_{f, x} \sin{\delta} - l_r F_{r, y}.
+\end{aligned}
+$$
+</p>
+
+with
+
+<p>
+$$
+\begin{aligned}
+&\alpha_f = - \arctan\left(\frac{\omega l_f + v_y}{v_x}\right) + \delta, \\
+&\alpha_r = \arctan\left(\frac{\omega l_r - v_y}{v_x}\right), \\
+&F_{f, y} = D_f \sin(C_f \arctan(B_f \alpha_f)), \\
+&F_{r, y} = D_r \sin(C_r \arctan(B_r \alpha_r)), \\
+&F_{f, x} = F_{r, x} = F_x, \\
+&F_x = (C_{m1} - C_{m2} v_x)\tilde{d} - C_{m3} - C_{m4} v^2_x.
+\end{aligned}
+$$
+</p>
+
+where
+$m$ is the mass of the vehicle, $F_{r, x},\ F_{r, y},\ F_{f, x},\text{ and } F_{f, y} \in \R$ are the forces acting on the tires with $r$ and $f$ indicating rear and front tires respectively, and $x$ and $y$ indicating longitudinal and lateral component of forces in the local vehicle body xy-frame respectively. The constants $l_f,\text{ and }  l_r \in \R_+$ are the distances from the Centre of Gravity $(C_g)$ to front and rear wheels respectively. The control inputs $\tilde{d} \in \R_{[0, 1]},\text{ and }  \delta \in \R$ are the normalised forward acceleration PWM duty-cycle control input and front wheel steering angle input in radians respectively. The states $p_x,\ p_y,\text{ and }  \psi \in \R$ are the vehicle's x position in meters, the y position in meters, and the heading in radians in the global frame of reference respectively. The states $v_x,\ v_y,\text{ and }  \omega \in \R$ are the vehicle's x velocity in meters-per-second, the y velocity in meters-per-second, and the angular rotation rate around the z-axis in radians-per-second in the local body-fixed frame of reference respectively. The constants $B_f$, and $B_r \in \R$ are the *stiffness factors*, $C_f$, and $C_r \in \R$ are the *shape factors* $D_f$, and $D_r \in \R$ are the *peak factors*, and $C_{m1}$, $C_{m2}$, $C_{m3}$, and $C_{m4} \in \R_+$ are empirical parameters as described in [1]. The parameters $\alpha_f$ and $\alpha_r$ are the slip angles. These equations and their parameters are described in equations (1) through (4) in [1]. Now the dynamics can be written as,
+
+<p>
+$$
+\begin{aligned}
+&\dot{\mathbf{z}}(t) = \mathbf{g}(\mathbf{z}(t),\ \mathbf{u}(t)), \\
+\text{with}\\
+&\mathbf{z}(t) = [p_x(t),\ p_y(t),\ \psi(t),\ v_x(t),\ v_y(t),\ \omega(t)]^\mathsf{T} \text{ as the state vector}, \\
+&\mathbf{u}(t) = [\tilde{d}(t),\ \delta(t)]^\mathsf{T} \text{ as the input vector}, \\
+&\mathbf{g} : \R^{n_x} \times \R^{n_u} \rightarrow \R^{n_x}, \\
+&n_x = 6 \text{ and } n_u = 2.
+\end{aligned}
+$$
+</p>
+
+Using forward Euler method, the discretised system for some small sampling time $T \in \R_{>0}$ can be written as,
```

Comment:
$T \in R_{>0}$ change to $T>0$ (always choose the simplest possible notation)

Reply:
Done.